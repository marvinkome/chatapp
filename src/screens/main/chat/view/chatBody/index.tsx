import * as React from 'react';
import { Subscription } from 'react-apollo';

// UI Element
import ChatBody from './body';

// graphql documents
import { userTypingEvent } from './gql';

import { contextConnect } from '../../../../lib/context';
import context from '../../context';
import { ViewProps } from '../../types';

export interface Props {
    groupId: string;
    subscribe: () => void;
}

class MainChatBody extends React.Component<Props> {
    componentDidMount() {
        // subscribe to more messages
        this.props.subscribe();
    }

    render() {
        const subscriptionProps = {
            subscription: userTypingEvent,
            variables: { groupId: this.props.groupId }
        };

        return (
            <Subscription {...subscriptionProps}>
                {({ data }) => {
                    return <ChatBody typingData={data} {...this.props} />;
                }}
            </Subscription>
        );
    }
}

export default contextConnect(context, (store: ViewProps) => ({
    groupId: store.group.id,
    subscribe: store.subscribe
}))(MainChatBody);
