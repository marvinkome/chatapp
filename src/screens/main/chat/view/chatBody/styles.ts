import { StyleSheet } from 'react-native';
import theme from '../../../../../lib/colors';

export const chatMsg = StyleSheet.create({
    listContainer: {
        flex: 1,
        backgroundColor: 'transparent',
        borderBottomWidth: 0,
        borderTopWidth: 0,
        marginTop: 5
    },
    itemContainer: {
        borderBottomWidth: 0,
        paddingLeft: 5,
        paddingBottom: 20
    },
    container: {
        alignItems: 'flex-start'
    },
    titleStyle: {
        fontWeight: '500',
        color: theme.primary.typo.main,
        fontFamily: theme.fontRegular,
        fontSize: 16,
        marginLeft: 10
    },
    message: {
        color: theme.primary.typo.sub,
        fontSize: 14,
        marginTop: 2,
        marginLeft: 10,
        fontFamily: theme.fontRegular
    },
    timestamp: {
        color: '#86939e',
        fontSize: 12,
        marginTop: 1,
        marginLeft: 10,
        fontWeight: 'normal',
        fontFamily: theme.fontRegular
    },
    typingIndicator: {
        fontFamily: theme.fontRegular,
        color: theme.primary.typo.text,
        marginLeft: 20,
        fontSize: 12
    },
    dateDivider: {
        backgroundColor: theme.primary.light
    }
});
