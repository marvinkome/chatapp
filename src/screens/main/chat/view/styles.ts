import { StyleSheet } from 'react-native';
import theme from '../../../../lib/colors';

export const viewStyles = StyleSheet.create({
    container: {
        backgroundColor: theme.primary.regular,
        flex: 1
    }
});

export const chatHeaderStyles = StyleSheet.create({
    container: {
        backgroundColor: theme.primary.regular,
        elevation: 4,
        borderBottomWidth: 0
    },
    leftCont: {
        flexDirection: 'row',
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center'
    },
    icon: {
        fontSize: 37,
        color: theme.primary.typo.main,
        padding: 10,
        borderRadius: 5
    },
    rightIcon: {
        fontSize: 17,
        color: theme.primary.typo.main,
        paddingBottom: 15
    },
    title: {
        fontSize: 19,
        fontFamily: theme.fontRegular,
        paddingBottom: 14,
        paddingLeft: 15,
        color: theme.primary.typo.main
    },
    menuOptions: {
        paddingTop: 5,
        paddingBottom: 5
    },
    menuText: {
        fontFamily: 'PT_Sans',
        fontSize: 15,
        padding: 7,
        color: theme.text_title
    }
});

export const chatFormStyles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        height: 45
    },
    formContainer: {
        flex: 4,
        backgroundColor: theme.primary.regular,
        padding: 0,
        margin: 0
    },
    input: {
        color: theme.primary.typo.main,
        height: 45,
        width: '100%'
    },
    btnCont: {
        flex: 1
    },
    btn: {
        marginLeft: 0,
        marginRight: 0
    },
    btnStyle: {
        padding: 9,
        paddingLeft: 7,
        backgroundColor: theme.primary.regular
    },
    btnDisabled: {
        backgroundColor: theme.primary.regular
    },
    btnText: {
        display: 'none'
    },
    offlineBtn: {
        opacity: 0.5
    },
    offlineInput: {
        color: 'yellow',
        opacity: 0.5
    }
});
