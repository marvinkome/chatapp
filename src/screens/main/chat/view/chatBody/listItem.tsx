import * as React from 'react';
import { Text } from 'react-native';
import { ListItem, Avatar } from 'react-native-elements';
import { chatMsg as styles } from './styles';
import { formatDate } from '../../../../../lib/helpers';
import theme from '../../../../../lib/colors';

interface Props {
    id: string;
    from: {
        id: string;
        username: string;
        profilePic: string;
    };
    message: string;
    timestamp: string;
}

const isError = (id: number) => -100 > id && id > -200;
const isOptimistic = (id: number) => 0 > id && id > -100;

export function ChatMsg(props: Props) {
    const optimisticStyle = isOptimistic(Number(props.id))
        ? {
              backgroundColor: theme.primary.light
          }
        : {};

    const errorStyle = isError(Number(props.id))
        ? {
              color: theme.error.bg
          }
        : {};

    const subtitle = (
        <React.Fragment>
            <Text style={[styles.message, errorStyle]}>{props.message}</Text>
            {isError(Number(props.id)) && (
                <Text
                    style={[
                        styles.message,
                        errorStyle,
                        {
                            fontSize: 12
                        }
                    ]}
                >
                    Can't send message at ths time
                </Text>
            )}
        </React.Fragment>
    );
    const title = (
        <Text style={styles.titleStyle}>
            {props.from.username}{' '}
            <Text style={styles.timestamp}>{formatDate(Number(props.timestamp), true)}</Text>
        </Text>
    );

    const avatarProps = {
        ...(!props.from.profilePic
            ? { title: props.from.username.slice(0, 2).toUpperCase() }
            : { source: { uri: props.from.profilePic } })
    };

    return (
        <ListItem
            title={title}
            subtitle={subtitle}
            hideChevron
            avatar={<Avatar rounded {...avatarProps} />}
            containerStyle={[styles.itemContainer, optimisticStyle]}
            wrapperStyle={styles.container}
            titleStyle={styles.titleStyle}
        />
    );
}
