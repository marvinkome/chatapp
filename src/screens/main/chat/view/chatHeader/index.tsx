import * as React from 'react';
import { Text, TouchableNativeFeedback } from 'react-native';
import { withNavigation, NavigationInjectedProps } from 'react-navigation';
import { Header, Icon } from 'react-native-elements';

import RightComponent from './rightIcon';
import { chatHeaderStyles as styles } from '../styles';

import { contextConnect } from '../../../../../lib/context';
import context from '../../context';
import { ViewProps } from '../../types';

// TODO: add wave effect
const LeftIcon = withNavigation((props: NavigationInjectedProps) => (
    <TouchableNativeFeedback onPress={() => props.navigation.goBack()}>
        <Icon name="ios-arrow-round-back" type="ionicon" iconStyle={styles.icon} />
    </TouchableNativeFeedback>
));

const CenterComponent = ({ name }: { name: string }) => (
    <Text data-testID="username" style={styles.title}>
        {name}
    </Text>
);

interface Props {
    name: string;
    role: string;
    groupId: string;
}
export const ChatHeader = (props: Props) => {
    return (
        <Header
            outerContainerStyles={styles.container}
            // @ts-ignore
            leftComponent={<LeftIcon />}
            centerComponent={<CenterComponent name={props.name} />}
            rightComponent={<RightComponent role={props.role} groupId={props.groupId} />}
        />
    );
};

function mapContextToProps(store: ViewProps) {
    return {
        name: store.group.name,
        role: store.group.role.name,
        groupId: store.group.id
    };
}

export default contextConnect(context, mapContextToProps)(ChatHeader);
