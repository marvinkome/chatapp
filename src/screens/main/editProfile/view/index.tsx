import * as React from 'react';
import { View, ScrollView } from 'react-native';
import { Avatar } from 'react-native-elements';
import { uploadImage, showImagePicker } from '../../../../lib/helpers';
import Loader from '../../../../components/loader';

import { viewStyle as styles } from './styles';
import { LinkForm } from './form';

interface IProps {
    user: {
        id: string;
        username: string;
        about: string;
        email: string;
        profilePic: string;
    };
    updateUser: (value: any) => void;
}

export default class PageView extends React.Component<IProps> {
    state = {
        imageUrl: '',
        loading: false
    };

    showPicker = async () => {
        try {
            const data = await showImagePicker('Select profile picture');
            const imageUrl = await uploadImage(data.path, data.name);
            if (!imageUrl) {
                return;
            }

            this.setState({ imageUrl });
        } catch (e) {
            return;
        }
    };

    updateUser = async (user: any) => {
        this.setState({ loading: true });
        await this.props.updateUser({ ...user, ...this.state });
        this.setState({ loading: false });
    };

    render() {
        const avatarProps = {
            ...(!this.props.user.profilePic
                ? {
                      ...(this.state.imageUrl.length
                          ? { source: { uri: this.state.imageUrl } }
                          : { title: this.props.user.username.slice(0, 2).toUpperCase() })
                  }
                : { source: { uri: this.props.user.profilePic } })
        };

        return (
            <ScrollView style={styles.container}>
                <View style={styles.avatar}>
                    <Avatar rounded large onPress={this.showPicker} {...avatarProps} />
                </View>
                <LinkForm user={this.props.user} updateUser={this.updateUser} />

                {this.state.loading && <Loader translucent />}
            </ScrollView>
        );
    }
}
