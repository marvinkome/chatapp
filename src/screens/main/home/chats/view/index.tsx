import * as React from 'react';

// main imports
import { compose, ComponentEnhancer as CE } from 'recompose';
import orderBy from 'lodash/orderBy';
import { withNavigation, NavigationInjectedProps as N } from 'react-navigation';
import firebase from 'react-native-firebase';

// UI
import { View, FlatList } from 'react-native';
import { List } from 'react-native-elements';

// components
import ListItem from './listing';
import { GroupFab } from './components/fab';
import { EmptyList } from './components/emptyMsg';

// styles
import { ViewStyles as styles } from './styles';

// types
import { ListingType, ViewProps } from '../types';

// helpers
import { formatDate } from '../../../../../lib/helpers';
const image = require('../../../../../../static/pp.jpg');

type ViewState = {
    usersTyping: object;
};

class MainView extends React.Component<N & ViewProps & {}, ViewState> {
    state: ViewState = {
        usersTyping: {}
    };

    componentDidMount() {
        this.subscribeToGroup();
    }
    componentDidUpdate(prevProps: ViewProps) {
        // if no new data
        if (!this.props.typingData) {
            return;
        }

        // if prev props is falsy then return new state
        if (!prevProps.typingData) {
            const ut = this.updateTypingUsers(this.props.typingData);
            return this.setState({
                usersTyping: ut
            });
        }

        // if isTyping or username is un-changed
        const { userGroupTyping: oldData } = prevProps.typingData;
        const { userGroupTyping: newData } = this.props.typingData;

        if (
            oldData.isTyping === newData.isTyping &&
            oldData.user.username === newData.user.username
        ) {
            return;
        }

        // when theres no problem
        const usersTyping = this.updateTypingUsers(this.props.typingData);
        return this.setState({
            usersTyping
        });
    }

    updateTypingUsers = (typingData: any) => {
        // check that typingData is available
        const { usersTyping } = this.state;

        const {
            isTyping,
            group: { id },
            user: { username }
        } = typingData.userGroupTyping;

        if (isTyping) {
            // if typing state is true
            if (usersTyping[id]) {
                // if group data is in state
                // add user to group in local state
                if (usersTyping[id].indexOf(username) >= 0) {
                    // if user is already in group
                    return usersTyping; // return the current state
                } else {
                    return {
                        ...usersTyping,
                        [id]: usersTyping[id].concat(username)
                    }; // add user to state
                }
            } else {
                return {
                    ...usersTyping,
                    [id]: [username]
                }; // add new group with user to state
            }
        } else {
            // remove user from local state
            return {
                ...usersTyping,
                ...(usersTyping[id] && {
                    [id]: usersTyping[id].filter((item: string) => item !== username)
                })
            };
        }
    };

    getTypingUsers(groupID: string) {
        const users = this.state.usersTyping[groupID];

        if (!users) {
            return null;
        }

        const length = users.length;

        if (!length) {
            return null;
        }

        const text = length === 1 ? 'is' : 'are';
        const punc = length === 1 ? '' : ',';
        const multiple = length > 2;

        const typingUsers = multiple
            ? 'multiple people'
            : users.join(`${punc} `).replace(/,([^,]*)$/, ' and$1');

        return `${typingUsers} ${text} typing...`;
    }

    subscribeToGroup() {
        this.props.data.groups.forEach((group) => {
            firebase.messaging().subscribeToTopic(group.id);
        });
    }

    formatItem = () => {
        const groups = this.props.data.groups;

        const formattedGroups = groups.reduce((reduced: ListingType[], group) => {
            // get icon
            const groupImage = group.image
                ? { source: { uri: group.image } }
                : { icon: { name: 'group', type: 'material-icons' } };

            // reduce item to match list props
            const unread = group.unreadCount;
            let item = {
                id: group.id,
                name: group.name,
                text: 'No message',
                timestamp: '',
                unread,
                image: groupImage,
                onPress: () => false,
                typing: false
            };

            // check if last message is available
            if (group.lastMessage) {
                item = {
                    ...item,
                    text: group.lastMessage.message,
                    timestamp: formatDate(Number(group.lastMessage.timestamp)),
                    onPress: () =>
                        this.props.navigation.navigate('Chat', {
                            groupId: group.id,
                            lastMessageId: group.lastMessage.timestamp
                        })
                };
            }

            if (this.state.usersTyping[group.id] && this.state.usersTyping[group.id].length) {
                const typingUser = this.getTypingUsers(group.id);
                item = {
                    ...item,
                    ...(typingUser && {
                        text: typingUser,
                        typing: true
                    })
                };
            }

            reduced.push(item);
            return reduced;
        }, []);

        return orderBy(formattedGroups, 'timestamp', 'desc');
    };

    render() {
        const lists = this.formatItem();

        return (
            <View style={styles.container}>
                {/* check length of groups */}
                {this.props.data.groups.length ? (
                    <List containerStyle={styles.listContainer}>
                        <FlatList
                            data={lists}
                            keyExtractor={(item) => item.id}
                            data-testid="chat-list"
                            renderItem={({ item }) => <ListItem listItem={item} typing={false} />}
                        />
                    </List>
                ) : (
                    <EmptyList />
                )}

                {/* Fab to join and create group */}
                <GroupFab navigate={this.props.navigation.navigate} />
            </View>
        );
    }
}

const enhance: CE<{}, ViewProps> = compose(withNavigation);

export default enhance(MainView);
