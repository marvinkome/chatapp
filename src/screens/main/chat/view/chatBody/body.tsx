import * as React from 'react';
import _ from 'lodash';
import moment from 'moment';

// UI Elements
import { FlatList, Text } from 'react-native';
import { List } from 'react-native-elements';
import Hr from '../../../../../components/Hr';
import { ChatMsg } from './listItem';

// styles
import theme from '../../../../../lib/colors';
import { chatMsg as style } from './styles';

// types
import { Props } from './index';

import { contextConnect } from '../../../../../lib/context';
import context from '../../context';
import { ViewProps } from '../../types';

const ITEM_HEIGHT = 72.667;
const THRESHOLD = 1;

type ChatBodyProps = Props & {
    items: Array<{
        id: string;
        from: {
            id: string;
            username: string;
        };
        message: string;
        timestamp: any;
    }>;
    typingData?: {
        userTyping: {
            user: {
                username: string;
            };
            isTyping: boolean;
        };
    };
    unreadCount: number;
    lastViewedMessage: string;
    fetchMore: () => void;
    refreshing: boolean;
};

type ChatBodyState = {
    usersTyping: string[];
};

export class ChatBody extends React.Component<ChatBodyProps, ChatBodyState> {
    state: ChatBodyState = {
        usersTyping: []
    };

    componentDidUpdate(prevProps: ChatBodyProps) {
        // if no new data
        if (!this.props.typingData) {
            return;
        }

        // if prev props is falsy then return new state
        if (!prevProps.typingData) {
            const ut = this.updateTypingUsers(this.props.typingData);
            return this.setState({
                usersTyping: ut
            });
        }

        // if isTyping or username is un-changed
        const { userTyping: oldData } = prevProps.typingData;
        const { userTyping: newData } = this.props.typingData;

        if (
            oldData.isTyping === newData.isTyping &&
            oldData.user.username === newData.user.username
        ) {
            return;
        }

        // when theres no problem
        const usersTyping = this.updateTypingUsers(this.props.typingData);
        return this.setState({
            usersTyping
        });
    }

    updateTypingUsers = (typingData: any) => {
        // check that typingData is available
        const { usersTyping } = this.state;

        const {
            isTyping,
            user: { username }
        } = typingData.userTyping;

        // check if typing state is true
        if (isTyping) {
            // add user to local state
            if (usersTyping.indexOf(username) >= 0) {
                // check if user is not already in state
                return usersTyping; // return the current state
            } else {
                return usersTyping.concat(username); // add user to state
            }
        } else {
            // remove user from local state
            return usersTyping.filter((item) => item !== username);
        }
    };

    checkIfBeginningReached = ({ nativeEvent }: any) => {
        const { layoutMeasurement, contentOffset } = nativeEvent;
        const currentPos = layoutMeasurement.height + contentOffset.y;
        const listLength = ITEM_HEIGHT * this.props.items.length;
        const reactThreshold = listLength - ITEM_HEIGHT * THRESHOLD;

        if (!this.props.refreshing && currentPos >= reactThreshold) {
            this.props.fetchMore();
        }
    };

    renderItem = (props: any) => {
        const newMessageId = this.props.lastViewedMessage;

        return props.item.messages.map((item: any) => {
            return (
                <React.Fragment key={item.id}>
                    <ChatMsg {...item} />
                    {newMessageId &&
                        item.id === newMessageId && (
                            <Hr
                                lineStyle={{ backgroundColor: theme.secondary.regular }}
                                textStyle={{ color: theme.secondary.typo.sub }}
                                text="New Message"
                            />
                        )}
                </React.Fragment>
            );
        });
    };

    renderTypingUsers() {
        const users = this.state.usersTyping;
        const length = users.length;

        if (!length) {
            return null;
        }

        const text = length === 1 ? 'is' : 'are';
        const punc = length === 1 ? '' : ',';
        const multiple = length > 3;

        return (
            <Text style={style.typingIndicator}>
                {multiple
                    ? 'multiple people'
                    : users.join(`${punc} `).replace(/,([^,]*)$/, ' and$1')}{' '}
                {text} typing...
            </Text>
        );
    }

    renderEmptyGroup() {
        return <Text style={style.emptyText}>Welcome to the beginning of the group</Text>;
    }

    itemSeparator = ({ leadingItem }: any) => (
        <Hr
            text={moment(leadingItem.timestamp, 'ddd MMM DD YYYY').format('D MMM Y')}
            lineStyle={style.dateDivider}
        />
    );

    render() {
        const data = _.chain(this.props.items)
            .groupBy((item) => new Date(Number(item.timestamp)).toDateString())
            .map((messages, timestamp) => ({ messages, timestamp }))
            .value();

        return (
            <List containerStyle={style.listContainer}>
                <FlatList
                    data={data}
                    renderItem={this.renderItem}
                    ItemSeparatorComponent={this.itemSeparator}
                    keyExtractor={(item) => item.timestamp}
                    onScroll={this.checkIfBeginningReached}
                    ListEmptyComponent={this.renderEmptyGroup()}
                    inverted
                />
                {this.renderTypingUsers()}
            </List>
        );
    }
}

const mapper = ({ group, fetchMore, refreshing }: ViewProps) => ({
    items: group.messages,
    groupId: group.id,
    unreadCount: group.unreadCount,
    lastViewedMessage: group.lastViewedMessage,
    fetchMore,
    refreshing
});

export default contextConnect(context, mapper)(ChatBody);
