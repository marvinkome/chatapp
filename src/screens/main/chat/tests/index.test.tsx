import React from 'react';
import Chats from '../index';
import View from '../view';
import { ChatBody } from '../view/chatBody';
import { ChatForm } from '../view/chatForm';
import { create } from 'react-test-renderer';

jest.mock('socket.io-client');

describe('chat screen', () => {
    it('renders correctly', () => {
        const comp = create(<Chats />);
        expect(comp.toJSON()).toMatchSnapshot();
    });

    describe('chat view', () => {
        const view = create(<View />);
        const viewIns = view.root;
        it('renders (snapshot)', () => {
            expect(view.toJSON()).toMatchSnapshot();
        });

        // should containing the to - user name
        it('should contain the reciever username and last seen', () => {
            const comp = viewIns.findByProps({
                ['data-testId']: 'chat-header'
            });
            const name = comp.findByProps({ ['data-testID']: 'username' });
            const lastSeen = comp.findByProps({ ['data-testID']: 'last-seen' });

            expect(name.props.children).toBe('Larry Doe');
            expect(lastSeen.props.children).toBe('Last seen: 12/12/17');
        });

        describe('chat body', () => {
            const listItems = [
                {
                    name: 'jamesbond',
                    text: 'Lorem ipsum dolor sit amet consectetur',
                    time: Date.now()
                },
                {
                    name: 'you',
                    text: 'Doing?',
                    time: Date.now()
                }
            ];
            it('renders 4 chats', () => {
                const chatItems = create(<ChatBody items={listItems} />);
                expect(chatItems).toMatchSnapshot();
            });

            // should have avatar component only on messages from sender
            // only first message should have avatar image when there's multple messages
            // there should be a new messages indicator
        });

        describe('chat form', () => {
            it('should render correctly', () => {
                const comp = create(<ChatForm sendMessage={jest.fn()} />);
                expect(comp).toMatchSnapshot();
            });
        });
    });
});
