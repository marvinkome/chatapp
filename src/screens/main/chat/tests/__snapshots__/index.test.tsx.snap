// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat screen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "hsl(0, 0%, 96%)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#476DC5",
          "borderBottomColor": "#f2f2f2",
          "borderBottomWidth": 1,
          "height": 70,
          "padding": 15,
        },
        undefined,
        Object {
          "backgroundColor": "#fff",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <View>
        <View
          style={
            Array [
              false,
              false,
              false,
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          underlayColor="white"
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "black",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Object {
                    "color": "#42ED86",
                    "fontSize": 37,
                  },
                ],
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          data-testID="username"
          ellipsizeMode="tail"
          style={
            Object {
              "color": "hsl(0, 0%, 20%)",
              "fontFamily": "Karla",
              "fontSize": 19,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Larry Doe
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          data-testID="last-seen"
          ellipsizeMode="tail"
          style={
            Object {
              "color": "hsl(0, 0%, 45%)",
              "fontFamily": "Karla",
              "fontSize": 14,
              "textAlign": "center",
            }
          }
        >
          Last seen: 12/12/17
        </Text>
      </View>
      <View
        style={
          Object {
            "height": 24,
            "width": 24,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    onContentSizeChange={[Function]}
    style={
      Object {
        "flex": 1,
        "padding": 16,
      }
    }
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 4,
        }
      }
    >
      <View
        style={
          Object {
            "borderBottomColor": "#bdc6cf",
            "borderBottomWidth": 1,
            "marginLeft": 20,
            "marginRight": 20,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={true}
          onChangeText={[Function]}
          placeholder="Write a message..."
          style={
            Array [
              Object {
                "color": "#86939e",
                "minHeight": 36,
                "width": 750,
              },
              Object {
                "fontSize": 17.5,
              },
              Object {
                "height": 65,
              },
            ]
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "marginLeft": 15,
              "marginRight": 15,
            },
            undefined,
            Object {
              "marginLeft": 0,
              "marginRight": 0,
            },
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            pointerEvents="box-only"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#9E9E9E",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 19,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                false,
                undefined,
                undefined,
                undefined,
                Object {
                  "backgroundColor": "#42ED86",
                  "paddingLeft": 7,
                },
                undefined,
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 20,
                    },
                    undefined,
                    false,
                    undefined,
                    Object {
                      "display": "none",
                    },
                    undefined,
                    undefined,
                    undefined,
                  ],
                ]
              }
            >
              BTN
            </Text>
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 26,
                  },
                  Array [
                    Object {
                      "marginLeft": 10,
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              type="ionicon"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
